File: TwoPointer/2116.check-if-a-parentheses-string-can-be-valid.py
# @leet start
class Solution:
    def canBeValid(self, s: str, locked: str) -> bool:
        if len(s) % 2 == 1:
            return False
        sema = 0
        for i in range(len(s)):
            if s[i] == "(" or locked[i] == "0":
                sema += 1
            else:
                sema -= 1
            if sema < 0:
                return False
        sema = 0
        for i in range(len(s) - 1, -1, -1):
            if s[i] == ")" or locked[i] == "0":
                sema += 1
            else:
                sema -= 1
            if sema < 0:
                return False
        return True


# @leet end

File: TwoPointer/2116.check-if-a-parentheses-string-can-be-valid.go
package leetcode

// @leet start
func canBeValid(s string, locked string) bool {
	if len(s)%2 == 1 {
		return false
	}
	var sema int
	for i, r := range s {
		if locked[i] == '0' || r == '(' {
			sema++
		} else if r == ')' {
			sema--
		}
		if sema < 0 {
			return false
		}
	}
	sema = 0
	for i := len(s) - 1; i >= 0; i-- {
		r := s[i]
		if locked[i] == '0' || r == ')' {
			sema++
		} else if r == '(' {
			sema--
		}
		if sema < 0 {
			return false
		}
	}
	return true
}

// @leet end

